### PR TITLE
WIP: Removal of identity nodes in matrix DDs

### PR DESCRIPTION
**Pull request by aaronleesander**
This pull request contains the changes which remove identity nodes in matrix DDs.

This includes not constructing a gate DD for the full system size (i.e. a single qubit operation is expanded to n-qubit with identities) as well as ignoring/removing identities in operations such as multiplication.

### CODE REVIEW

File: include/dd/Package.hpp [hunk 1]
```diff
@@ -579,11 +575,11 @@ namespace dd {
             auto       z             = static_cast<Qubit>(start);
             const auto smallerTarget = std::min(target0, target1);
             for (; z < smallerTarget; ++z) {
```
Comment (burgholzer) [inline include/dd/Package.hpp:577]: This whole function can be simplified quite a lot. As there are no controls, it simply constructs a two-level DD with the appropriate nodes. All the special handling below, in-between, and above targets can be eliminated.

Actually, at that point (also relating to the comment above), this basically becomes the routine that takes a matrix and constructs the DD. Maybe that can also be reused here or used instead of the method here.

File: include/dd/Package.hpp [hunk 2]
```diff
@@ -508,8 +508,6 @@ namespace dd {
                             } else { // pos. control
                                 em[i] = makeDDNode(z, std::array{(i1 == i2) ? makeIdent(static_cast<Qubit>(start), static_cast<Qubit>(z - 1)) : mEdge::zero, mEdge::zero, mEdge::zero, em[i]});
                             }
-                        } else { // not connected
```
Comment (burgholzer) [inline include/dd/Package.hpp:511]: It could be worth to introduce a separate routine that handles the no-control case as this becomes almost trivial now. Technically one could use the routine that takes a matrix and creates a DD out of that.

The function here could then become something like
```c++
if (controls.empty()) {
 return ... [call special function for just constructing the single qubit operation matrix] ...
}

... [do the usual] ...
```

File: include/dd/Package.hpp [hunk 3]
```diff
@@ -1007,7 +1003,7 @@ namespace dd {
             assert(e.p->ref == 0);
             for ([[maybe_unused]] const auto& edge: edges) {
                 // an error here indicates that cached nodes are assigned multiple times. Check if garbage collect correctly resets the cache tables!
-                assert(edge.p->v == var - 1 || edge.isTerminal());
+                // assert(edge.p->v == var - 1 || edge.isTerminal());
```
Comment (burgholzer) [inline include/dd/Package.hpp:1006]: Don't just comment out stuff. Actually remove it ;o)

File: include/dd/Package.hpp [hunk 4]
```diff
@@ -1662,6 +1658,22 @@ namespace dd {
...
+                        e = yCopy;
+                    }
+                    computeTable.insert(xCopy, yCopy, {e.p, e.w});
+                    e.w = cn.mulCached(x.w, y.w);
+
+                    if (e.w.approximatelyZero()) {
+                        cn.returnToCache(e.w);
+                        return ResultEdge::zero;
+                    }
+                    return e;
```
Comment (burgholzer) [inline include/dd/Package.hpp:1676]: this seems wrong to me. Just because there is an identity at the current level does not mean that the result of the multiplication is just the other DD. This should still require recursive calls to properly handle the multiplication.

File: include/dd/Package.hpp [hunk 5]
```diff
@@ -1675,17 +1687,27 @@ namespace dd {
...
                     for (auto k = 0U; k < rows; k++) {
                         LEdge e1{};
+                        auto  element1 = rows * i + k;
                         if (!x.isTerminal() && x.p->v == var) {
-                            e1 = x.p->e[rows * i + k];
+                            e1 = x.p->e[element1];
                         } else {
+                            // Effectively inserts an identity node
                             e1 = xCopy;
+                            if ((element1 == 1 or element1 == 2) and (std::is_same_v<LeftOperandNode, mNode> or std::is_same_v<LeftOperandNode, dNode>)) {
```
Comment (burgholzer) [inline include/dd/Package.hpp:1696]: Please do not use logical connectives like `and` or `or`. There are many instances of this leading to problems.
Prefer the standard notation of `&&` and `||`.

File: include/dd/Package.hpp [hunk 6]
```diff
@@ -1675,17 +1687,27 @@ namespace dd {
                     edge[idx] = ResultEdge::zero;
                     for (auto k = 0U; k < rows; k++) {
                         LEdge e1{};
+                        auto  element1 = rows * i + k;
```
Comment (burgholzer) [inline include/dd/Package.hpp:1690]:
```suggestion
                        const auto  element1 = rows * i + k;
```

File: include/dd/Package.hpp [hunk 7]
```diff
@@ -1675,17 +1687,27 @@ namespace dd {
...
                         if (!x.isTerminal() && x.p->v == var) {
-                            e1 = x.p->e[rows * i + k];
+                            e1 = x.p->e[element1];
                         } else {
+                            // Effectively inserts an identity node
                             e1 = xCopy;
+                            if ((element1 == 1 or element1 == 2) and (std::is_same_v<LeftOperandNode, mNode> or std::is_same_v<LeftOperandNode, dNode>)) {
+                                e1.w = Complex::zero;
```
Comment (burgholzer) [inline include/dd/Package.hpp:1697]: I believe that this can be improved. This is not really generic and efficient. Same applied to the statement below:
```c++
if constexpr (std::is_same_v<LeftOperandNode, vNode>) {
  e1 = xCopy;
} else {
  if (i == k) {
    e1 = xCopy;
  } else {
    e1 = Ledge::zero;
}
```

Note that this could potentially be improved even further. As of now, The left operand node can never be a vector DD. That should probably be a `static_assert` somewhere within this method. So the first (constexpr) `if` condition is technically not even necessary here.
Furthermore, we don't support skipped nodes for vector DDs at the moment. This limitation could be made more explicit.

File: include/dd/Package.hpp
Comment (burgholzer) [inline include/dd/Package.hpp:1]: Is addition really already working? I would have guessed that changes are required there as well.
In general, I believe that there are many more routines that need adaptation and/or changes.
One particular example: Matrix nodes currently have a special flag (`isIdentity`) and a corresponding method, where this is set. This should no longer be necessary as all identities are removed and the only node that is now the identity is the one terminal (or, technically, a skipped node).

File: test/test_package.cpp
Comment (burgholzer) [inline test/test_package.cpp:1]: This will need quite some more tests on various functions to really make sure that everything works properly.

**Review by burgholzer — CHANGES_REQUESTED** [review]
Just a very quick review of the current changes. This is heading in the right direction, but will require quite some more changes to work properly.

File: test/test_package.cpp [hunk 8]
```diff
@@ -1804,3 +1794,29 @@ TEST(DDPackageTest, InnerProductTopNodeConjugation) {
...
+    auto newState = dd->multiply(hGate, zeroState);
+    dd->printVector(newState);
+
+    ASSERT_EQ(dd->getValueByPath(newState, 0), (dd::ComplexValue{dd::SQRT2_2, 0}));
+    ASSERT_EQ(dd->getValueByPath(newState, 1), (dd::ComplexValue{dd::SQRT2_2, 0}));
+    ASSERT_EQ(dd->getValueByPath(newState, 2), (dd::ComplexValue{0, 0}));
+    ASSERT_EQ(dd->getValueByPath(newState, 3), (dd::ComplexValue{0, 0}));
+
+    auto cxGate = dd->makeGateDD(dd::Xmat, 5, 0_pc, 1);
+    EXPECT_EQ(dd->size(cxGate), 4);
```
Comment (burgholzer) [inline test/test_package.cpp:1813]:
```suggestion
    EXPECT_EQ(dd->size(cxGate), 3);
```

Something I missed in the review. The CX gate should just require two nodes (plus the terminal).

Side note: we should probably adjust the size function so that it really only counts non-terminal nodes. This would be a breaking change though and should be done in a separate PR at some point.

File: include/dd/Package.hpp (hunk 1 above)
Comment (aaronleesander) [reply include/dd/Package.hpp]: I simplified the function like you mentioned, so now it only directly affects the targets.

I tried using the function makeDDFromMatrix to simplify it. First, I overloaded it to take the matrix of ComplexValue as input rather than just a CMat.

I'll leave this as a TODO for now, since it is currently working at least and slightly out of the scope of this PR.

File: include/dd/Package.hpp (hunk 2 above)
Comment (aaronleesander) [reply include/dd/Package.hpp:511]: I added an if statement to check if there are no controls as you said. This allows us to avoid the for loops over the levels of the DD for the control construction if there are none. No need for an extra function.

This issue is probably resolved.

File: include/dd/Package.hpp [hunk 9]
```diff
@@ -1626,13 +1641,15 @@ namespace dd {
...
+                std::cout << "Cached entered \n";
                 // This branch is only taken for matrices
                 if (x.p->v == var && x.p->v == y.p->v) {
+                    std::cout << "Cached if entered \n";
                     if (x.p->isIdentity()) {
                         if constexpr (n == NEDGE) {
                             // additionally check if y is the identity in case of matrix multiplication
                             if (y.p->isIdentity()) {
-                                e = makeIdent(start, var);
+                                // e = makeIdent(start, var);
```
Comment (burgholzer) [inline include/dd/Package.hpp:1652]: I think you need to be careful here when commenting this method out.
if `x` is the identity and `y` is the identity, then `e` is not properly set since you commented out the identity code. In fact, it isn't set at all.

File: include/dd/Package.hpp [hunk 10]
```diff
@@ -1662,6 +1679,23 @@ namespace dd {
...
+                    }
+                    computeTable.insert(xCopy, yCopy, {e.p, e.w});
+                    e.w = cn.mulCached(x.w, y.w);
+
+                    if (e.w.approximatelyZero()) {
+                        cn.returnToCache(e.w);
+                        return ResultEdge::zero;
+                    }
+                    std::cout << "Cached else entered \n";
+                    return e;
```
Comment (burgholzer) [inline include/dd/Package.hpp:1698]: First of all, this seems like duplicate code to me (see a few lines above). So consider combining these parts of the code.
Secondly, I believe one needs to take extra care, to do the right thing here. I'd suggest revisiting the whole logic of this `if ... else` construct here, as there can be no case where `x.p->v == y.p->v` and one of the nodes being an identity (since identities are removed entirely as part of this PR).

One possible way forward could be to just remove the `makeIdent` function, the `isIdentity` method of the Node class, as well as the `identity` flag itself from that class and see in which places the code breaks at the moment. Then, fix the corresponding places by either removing the code if it no longer applies, or adjust accordingly for the new DD structure.

**Comment by burgholzer** [comment]
[MQT Core](https://github.com/cda-tum/mqt-core) is now live and the DD package has been migrated there.
Please integrate your changes in a new PR to that repository. Since all files were migrated as-is, this should be straight-forward. Sorry for the inconvenience.
